resource-managers/kubernetes/core/src/main/scala/org/apache/spark/scheduler/cluster/k8s/ExecutorPodsLifecycleEventHandler.scala

```diff
@@ -0,0 +1,130 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.spark.scheduler.cluster.k8s
+
+import com.google.common.cache.{Cache, CacheBuilder}
+import io.fabric8.kubernetes.api.model.Pod
+import io.fabric8.kubernetes.client.KubernetesClient
+import scala.collection.JavaConverters._
+
+import org.apache.spark.SparkConf
+import org.apache.spark.deploy.k8s.Config._
+import org.apache.spark.deploy.k8s.Constants._
+import org.apache.spark.scheduler.ExecutorExited
+import org.apache.spark.util.Utils
+
+private[spark] class ExecutorPodsLifecycleEventHandler(
+    conf: SparkConf,
+    executorBuilder: KubernetesExecutorBuilder,
+    kubernetesClient: KubernetesClient,
+    podsEventQueue: ExecutorPodsEventQueue,
+    // Use a best-effort to track which executors have been removed already. It's not generally
+    // job-breaking if we remove executors more than once but it's ideal if we make an attempt
+    // to avoid doing so. Expire cache entries so that this data structure doesn't grow beyond
+    // bounds.
+    removedExecutorsCache: Cache[java.lang.Long, java.lang.Long]) {
+
+  import ExecutorPodsLifecycleEventHandler._
+
+  private val eventProcessingInterval = conf.get(KUBERNETES_EXECUTOR_EVENT_PROCESSING_INTERVAL)
+
+  def start(schedulerBackend: KubernetesClusterSchedulerBackend): Unit = {
+    podsEventQueue.addSubscriber(eventProcessingInterval) { updatedPods =>
+      updatedPods.foreach { updatedPod =>
+        processUpdatedPod(schedulerBackend, updatedPod)
+      }
+    }
+  }
+
+  private def processUpdatedPod(
+      schedulerBackend: KubernetesClusterSchedulerBackend, updatedPod: Pod) = {
+    val execId = updatedPod.getMetadata.getLabels.get(SPARK_EXECUTOR_ID_LABEL).toLong
+    if (isDeleted(updatedPod)) {
+      removeExecutorFromSpark(schedulerBackend, updatedPod, execId)
```

Comment:
we don't check for this in `ExecutorPodsAllocator`. So it could happen that we remove an executor here but not there.

Reply:
also both classes are very similar and have duplicated code. Can we factor out the commonalities? We could have a common base class that does the matching of pod phase and then calls the appropriate abstract method `handleFailurePhase`, `handleRunningPhase`, ..., which needs to be overwritten in the subclass.

Reply:
It's not entirely analogous but there are parts that can be shared. The pods allocator has a "post-batch" processing task which can run even on empty batches (particularly important for the first empty batch to request the first round of executors). But, I have some ideas for shared code that can be posted shortly.